contrib/pkg/awsprivatelink/endpointvpc/add.go

```diff
@@ -0,0 +1,466 @@
+	endpointVpcRegion string
+	endpointVpcId     string
+	endpointSubnetIds []string
+
+	dynamicClient             client.Client
+	endpointVpcClients        awsclient.Client
+	associatedVpcsWithClients []*awsutils.AssociatedVpcWithClient
+}
+
+func NewEndpointVPCAddCommand() *cobra.Command {
+	opt := &endpointVPCAddOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "add vpc-id",
+		Short: "Networking setup between an endpoint VPC and each associated VPC",
+		Long: `Networking setup between an endpoint VPC and each associated VPC 
+specified in HiveConfig.spec.awsPrivateLink.associatedVPCs: 
+1) Establish VPC peering connection between the endpoint VPC and the associated VPC
+2) Add a route (to the peered VPC) in relevant route tables of the associated VPC and the endpoint VPC 
+3) Update relevant security groups of the associated VPC and the endpoint VPC to allow traffic between them
+4) Add the endpoint VPC to HiveConfig.spec.awsPrivateLink.endpointVPCInventory`,
+		Args: cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := opt.Complete(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Validate(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Run(cmd, args); err != nil {
+				return
+			}
+		},
+	}
+
+	regionFlag := "region"
+	subnetIdsFlag := "subnet-ids"
+
+	flags := cmd.Flags()
+	flags.StringVar(&opt.endpointVpcRegion, regionFlag, "", "AWS Region of the endpoint VPC to add")
+	flags.StringSliceVar(&opt.endpointSubnetIds, subnetIdsFlag, []string{}, "IDs of the endpoint subnets to use (as a comma-separated string)")
+
+	_ = cmd.MarkFlagRequired(regionFlag)
+	_ = cmd.MarkFlagRequired(subnetIdsFlag)
+	return cmd
+}
+
+func (o *endpointVPCAddOptions) Complete(cmd *cobra.Command, args []string) error {
+	o.endpointVpcId = args[0]
+
+	// Get controller-runtime dynamic client
+	dynamicClient, err := hiveutils.GetClient()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create controller-runtime client")
+	}
+	o.dynamicClient = dynamicClient
+
+	// Get HiveConfig
+	hiveConfig := &hivev1.HiveConfig{}
+	if err = o.dynamicClient.Get(context.Background(), types.NamespacedName{Name: "hive"}, hiveConfig); err != nil {
+		log.WithError(err).Fatal("Failed to get HiveConfig/hive")
+	}
+
+	// Get associated VPCs and a set of AWS clients for each of them
+	associatedVpcsWithClients, err := awsutils.GetAssociatedVpcsWithClient(hiveConfig)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to retrieve associated VPCs and get AWS clients each of them")
+	}
+	if len(associatedVpcsWithClients) == 0 {
+		log.Warn("HiveConfig/hive does not specify any associated VPC. " +
+			"The endpoint VPC passed in as argument will still be added to HiveConfig, " +
+			"and yet no cloud resources will be set up for networking.")
+	}
+	o.associatedVpcsWithClients = associatedVpcsWithClients
+
+	// Get AWS clients for the endpoint VPC
+	awsClients, err := awsclient.NewClientFromSecret(nil, o.endpointVpcRegion)
```

Comment:
As below, this may end up being a redundant client if we already made one for this region.

Starting to think we should have `o.awsClientsByRegion[string]awsclient.Client` decoupled from the list of VPCs. Then if you like, you can have a little helper func like

```go
func (o *endpointVPCAddOptions) awsClientForVPC(vpc *hivev1.AWSPrivateLinkVPC) awsclient.Client {
	ret, ok := o.awsClientsByRegion[vpc.Region]
	if !ok {
		log.Fatalf("No AWS client for region %s", vpc.Region)
	}
	return ret
}
```

Reply:
Thanks for pointing that out. This is now implemented as follows (for both add/remove subcommands):

- First get a map of regions where we need AWS clients. This includes regions of the associated VPCs and the endpoint VPCs. 
- Then get a set of AWS clients for each region above.